akka-projection-cassandra/src/test/scala/akka/projection/cassandra/CassandraProjectionSpec.scala

```diff
@@ -0,0 +1,212 @@
+      }
+
+      val projection =
+        CassandraProjection
+          .atLeastOnce[Long, Envelope](projectionId, sourceProvider(entityId), offsetExtractor, 1, Duration.Zero) {
+            envelope =>
+              repository.concatToText(envelope.id, envelope.message)
+          }
+
+      projectionTestKit.run(projection) {
+        withClue("check - all values were concatenated") {
+          val concatStr = repository.findById(entityId).futureValue.get
+          concatStr.text shouldBe "abc|def|ghi|jkl|mno|pqr"
+        }
+      }
+      withClue("check - all offsets were seen") {
+        val offset = offsetStore.readOffset[Long](projectionId).futureValue.get
+        offset shouldBe 6L
+      }
+    }
+
+    "restart from previous offset - handler throwing an exception" in {
+      val entityId = UUID.randomUUID().toString
+      val projectionId = genRandomProjectionId()
+
+      val streamFailureMsg = "fail on fourth element"
+      val failingProjection =
+        CassandraProjection
+          .atLeastOnce[Long, Envelope](projectionId, sourceProvider(entityId), offsetExtractor, 1, Duration.Zero) {
+            envelope =>
+              if (envelope.offset == 4L) throw new RuntimeException(streamFailureMsg)
+              repository.concatToText(envelope.id, envelope.message)
+          }
+
+      withClue("check - offset is empty") {
+        val offsetOpt = offsetStore.readOffset[Long](projectionId).futureValue
+        offsetOpt shouldBe empty
+      }
+
+      val streamFailure =
+        intercept[RuntimeException] {
+          projectionTestKit.run(failingProjection) {
+            fail("fail assertFunc, we want to capture the stream failure")
+          }
+        }
+
+      withClue("check: projection failed with stream failure") {
+        streamFailure.getMessage shouldBe streamFailureMsg
+      }
+      withClue("check: projection is consumed up to third") {
+        val concatStr = repository.findById(entityId).futureValue.get
+        concatStr.text shouldBe "abc|def|ghi"
+      }
+      withClue("check: last seen offset is 3L") {
+        val offset = offsetStore.readOffset[Long](projectionId).futureValue.get
+        offset shouldBe 3L
+      }
+
+      // re-run projection without failing function
+      val projection =
+        CassandraProjection
+          .atLeastOnce[Long, Envelope](projectionId, sourceProvider(entityId), offsetExtractor, 1, Duration.Zero) {
+            envelope =>
+              repository.concatToText(envelope.id, envelope.message)
+          }
+
+      projectionTestKit.run(projection) {
+        withClue("checking: all values were concatenated") {
+          val concatStr = repository.findById(entityId).futureValue.get
+          concatStr.text shouldBe "abc|def|ghi|jkl|mno|pqr"
+        }
+      }
+
+      withClue("check: all offsets were seen") {
+        val offset = offsetStore.readOffset[Long](projectionId).futureValue.get
+        offset shouldBe 6L
+      }
+    }
+  }
```

Comment:
TODO: add test for saveOffsetAfterNumberOfEvents > 1 and saveOffsetAfterDuration > 0